internal/controller/controllers/agent_controller.go

```diff
@@ -211,6 +223,166 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, origAgent, &h.Host, h.ClusterID, nil, false)
 }
 
+func splitNamespacedName(nsName string) (string, string, error) {
+	parts := strings.Split(nsName, "/")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("unable to parse namespaced name %s", nsName)
+	}
+	return parts[0], parts[1], nil
+}
+
+func deleteBMHForMachine(ctx context.Context, spokeClient client.Client, machine *machinev1beta1.Machine) error {
+	bmhNSName, haveBMH := machine.GetAnnotations()["metal3.io/BareMetalHost"]
+	if !haveBMH {
+		return nil
+	}
+
+	bmhNamespace, bmhName, err := splitNamespacedName(bmhNSName)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse BMH name")
+	}
+
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+	bmhKey := client.ObjectKey{Namespace: bmhNamespace, Name: bmhName}
+	if err = spokeClient.Get(ctx, bmhKey, bmh); err != nil {
+		return errors.Wrapf(err, "failed to get BMH %s", bmhNSName)
+	}
+
+	// TODO: remove this once OCPBUGS-7581 is fixed
+	patch := client.MergeFrom(bmh.DeepCopy())
+	bmh.ObjectMeta.Finalizers = nil
+	if err = spokeClient.Patch(ctx, bmh, patch); err != nil {
+		return errors.Wrapf(err, "failed to remove BMH %s finalizers", bmhNSName)
+	}
+
+	if err = spokeClient.Delete(ctx, bmh); err != nil {
+		return errors.Wrapf(err, "failed to delete BMH %s", bmhNSName)
+	}
+
+	return nil
+}
+
+// removeSpokeResources removes all relevant resources from the agent's spoke cluster
+// This includes all or some of the node, machine, BMH, and scaling the machineset depending on what is present
+func (r *AgentReconciler) removeSpokeResources(ctx context.Context, log logrus.FieldLogger, agent *aiv1beta1.Agent) error {
+	spokeClient, err := r.spokeKubeClient(ctx, agent.Spec.ClusterDeploymentName)
+	if err != nil {
+		log.WithError(err).Error("failed to create spoke client")
+		return err
+	}
+
+	nodeName := getAgentHostname(agent)
+	log = log.WithField("node", nodeName)
+
+	nodeKey := client.ObjectKey{Name: nodeName}
+	node := &corev1.Node{}
+	if err = spokeClient.Get(ctx, nodeKey, node); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Warn("node not found")
+			return nil
+		}
+		log.WithError(err).Error("failed to get node from spoke cluster")
+		return err
+	}
+
+	deleteNode := func() error {
+		nodeErr := client.IgnoreNotFound(spokeClient.Delete(ctx, node))
+		if nodeErr != nil {
+			log.WithError(nodeErr).Error("failed to delete node")
+			return nodeErr
+		}
+		log.Info("spoke node deleted")
+		return nil
+	}
+
+	machineNSName, haveMachine := node.GetAnnotations()["machine.openshift.io/machine"]
+	if !haveMachine {
+		return deleteNode()
+	}
+
+	machineNamespace, machineName, err := splitNamespacedName(machineNSName)
+	if err != nil {
+		log.WithError(err).Error("failed to parse machine name from node annotation")
```

Comment:
It would help to log the value of the annotation here. In case this happens in the real world, it may be hard for our future selves to figure out what went wrong if we immediately deleted the input without preserving it somehow.

Reply:
Will do